retrieval/state.go

```diff
@@ -120,22 +132,90 @@ func (rs *State) GetPhase() RetrievalPhase {
 	return RetrievalPhase(rs.phase.Load())
 }
 
+// addProvider is a helper to add a provider to a sample list with size limit.
+// Only the first MaxProvidersSampleSize providers are kept to prevent unbounded memory growth.
+func (rs *State) addProvider(list *[]peer.ID, peerID peer.ID) {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	if len(*list) < MaxProvidersSampleSize {
+		*list = append(*list, peerID)
+	}
+}
```

Comment:
I think it would be more idiomatic if this looked more like `append`:
```suggestion
func (rs *State) appendProvider(list []peer.ID, peerIDs ...peer.ID) {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	if len(list) >= MaxProvidersSampleSize {
		return
	}
	if len(list) + len(peerIDs) > MaxProvidersSampleSize {
		peerIDs = peerIDs[:MaxProvidersSampleSize - len(list)]
	}
	return append(list, peerIDs...)
}
```